global_ids.md

```diff
@@ -0,0 +1,187 @@
+I think it'll probably make the most sense to consolidate commits into single
+documents in etcd. This has a number of benefits:
+
+- it should greatly simplify the code that keeps track of provenance and make
+  inconsistent states a lot harder to get into because consistency is defined
+  within a single document, rather than a chain of them.
+
+- It greatly decreases our etcd usage. In the current system a commit to a DAG
+  with 10 pipelines creates 11 commits in total. These are all committed
+  transactionally, and when deletions happen, deleted transactionally. Which
+  can get us over the op limit etcd pretty quickly. In this new system creating
+  and deleting a commit could only require a single op. It'll likely require
+  a bit more to adjust branches and commit parentage. But it'll be well
+  bounded, and not proportional to the size of the DAG.
+
+We can do a similar thing with jobs and move them into a single document for
+the whole DAG, and unlike with commits we kind of have to because jobs aren't
+namespaced the way that commits are. We also want to do this because it has
+similar benefits of decreasing the total number of documents we store in etcd.
+It should also simplify code a little, although the code for manipulating jobs
+is a lot simpler, so less to be gained there.
+
+## Complications
+While I think this overall simplifies our implementation it does introduce
+a few complications.
+
+### Stat Commits
+Stat commits are a little complicated because they occur in the same repo as
+the commits they're provenant on, which would cause a name collision. I think
+this is fairly easily patch by either namespacing commits by repo and branch,
+or by moving them into a separate repo.
+
+### Multiple inputs
+Pipelines with multiple inputs have create an interesting complication.
+Consider a DAG in which repos `A` and `B` are inputs to a pipeline `C`. When
+you commit to repo `A` you get a downstream commit in `C` with the same ID,
+however `B` doesn't have a copy of that commit now. So we're back to a world of
+needing to link commits with different IDs together to get reasonable
+provenance and subvenance. The solution is simply to have committing to `A`
+also commit to `B`, because they share subvenance. This might seem like it
+makes a lot of unnecessary commits, and in some sense it does, but in terms of
+what's stored in etcd it doesn't, since it just adds a few extra bytes to the
+commit document we're making anyway, which is less than the subvenance update
+we had to do in the old system. It does add an extra commit to the history of
+`B`, but I think that's overall a good thing, since it means you can use this
+ID to see which data in `B` was used to create the downstream commits. It's
+helpful to think of the new commit on `B` as not really a commit, as much as an
+alias for whatever's currently the head commit of master. That's basically how
+it will be implemented.
```

Comment:
For clarity, there _is_ a little bit of a decision here as to whether this is actually a new commit or just another ID that is an alias for an existing commit. Specifically, what is the parent of the new "alias" commit created on B. 

IMO and speaking from a what user might expect perspective, that commit is a figment of Pachyderm's lineage tracking, not a "real" commit and so therefore master^ should refer to the last "real" commit made in repo B. 

For example, user makes a single commit to A and B and then proceed to only ever commit to repo A. I see arguments both directions, but I think the most understandable notion is that there is really only ever 1 commit in repo B -- that commit is simply referencable by many different IDs. Therefore, there really is no master^ for repo B, but there is for A. Another reason why I think this logically makes sense is that if you look at most other tools in our space, they include a notion of "runs" (aka experiments in some systems) where a Run is basically a single runthrough of a whole DAG. Thinking of it that way, runs and commits are sort of separate ideas to users. In our system, the two are intertwined, which is fine, but it get's less understandible to users if there are actual commits being made. 

Another argument for this model, is imagine the same DAG but there is another pipeline with B as input only. If commits to A are actually making commits in B, as opposed to just creating an additional commit alias, that would seem to reason that the other pipeline that subscribes to B should be triggered too, but that seems pretty clearly incorrect. 

The biggest drawbacks of doing things this way are:
1. list commit get's a bit weird and we'd need to figure out exactly what we want that to mean. It's almost more like "list run", which isn't an actual noun in Pachyderm, unless we decide we want to make it one. 
2. master^ on repoA and repoB traverse commits, not IDs/runs. I think this is still preferable. In the scenario above of only having 1 commit in B, even though there are many IDs, there is actually only one version of the data. So, if a user trying to go to `B@master^`, that would fail because there is no parent and that's a good thing, it's giving the user meaningful information.

Reply:
I think to do this we'll want to introduce the notion of a "semantic parent." Parents have a lot of meaning behind them, including determining how processed datums get deduped. So I think putting another user constraint on the parent field might make things too complicated. But having the ability to traverse back to the parent that makes sense from the user's perspective will be good. I'm not sure what we'll want to do with `master^` I think we use that internally in some places, but we could move off of it. We may want to finally differentiate `master^` and `master~` like git does. Having one be the semantic parent and one the true parent.